pylint/checkers/typecheck.py

```diff
@@ -1811,25 +1814,24 @@ def _check_invalid_slice_index(self, node: nodes.Slice) -> None:
             if index is None:
                 continue
 
-            index_type = safe_infer(index)
-            if index_type is None or isinstance(index_type, util.UninferableBase):
-                continue
-
-            # Constants must be of type int or None
-            if isinstance(index_type, nodes.Const):
-                if isinstance(index_type.value, (int, type(None))):
-                    continue
-            # Instance values must be of type int, None or an object
-            # with __index__
-            elif isinstance(index_type, astroid.Instance):
-                if index_type.pytype() in {"builtins.int", "builtins.NoneType"}:
+            match index_type := safe_infer(index):
+                case _ if not index_type:
                     continue
+                case nodes.Const():
+                    # Constants must be of type int or None
+                    if isinstance(index_type.value, (int, type(None))):
```

Comment:
```suggestion
                case nodes.Const(value= int | type(None)):
```
Isn't this possible ?